src/mangrove/query_builder.hpp

```diff
@@ -1132,79 +1131,169 @@ class bit_update_expr {
 };
 
 /* Query comparison operators */
+// Note that each comparison operator has two versions:
+// a) One in which a field is compared to a value of the same type, and
+// b) One in which an array field is compared to a value of the contained type.
+//    (e.g. an array of int's can be compared to a single integer)
+//    This is syntactic sugar for MongoDB's $elemMatch operator, and mirrors
+//    the syntax in the actual query language.
 
 template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>>
 constexpr comparison_expr<NvpT, typename NvpT::no_opt_type> eq(
     const NvpT &lhs, const typename NvpT::no_opt_type &rhs) {
     return {lhs, rhs, "$eq"};
 }
 
+template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>,
+          typename = std::enable_if_t<is_iterable_v<typename NvpT::no_opt_type>>>
+constexpr comparison_expr<NvpT, typename NvpT::array_element_type> eq(
+    const NvpT &lhs, const typename NvpT::array_element_type &rhs) {
+    return {lhs, rhs, "$eq"};
+}
+
 template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>>
 constexpr comparison_expr<NvpT, typename NvpT::no_opt_type> operator==(
     const NvpT &lhs, const typename NvpT::no_opt_type &rhs) {
     return eq(lhs, rhs);
 }
 
+template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>,
+          typename = std::enable_if_t<is_iterable_v<typename NvpT::no_opt_type>>>
+constexpr comparison_expr<NvpT, typename NvpT::array_element_type> operator==(
+    const NvpT &lhs, const typename NvpT::array_element_type &rhs) {
+    return eq(lhs, rhs);
+}
+
 template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>>
 constexpr comparison_expr<NvpT, typename NvpT::no_opt_type> gt(
     const NvpT &lhs, const typename NvpT::no_opt_type &rhs) {
     return {lhs, rhs, "$gt"};
 }
 
+template <typename NvpT, typename = std::enable_if_t<is_nvp_v<NvpT>>,
+          typename = std::enable_if_t<is_iterable_v<typename NvpT::no_opt_type>>>
+constexpr comparison_expr<NvpT, typename NvpT::array_element_type> gt(
```

Comment:
I think these can go away using [std::conditional](http://en.cppreference.com/w/cpp/types/conditional)
Something like this(?):

```c++
template <
    typename NvpT,
    typename T = std::conditional_t<
        is_nvp_v<NvpT>,
        NvpT::no_opt_type,
        std::enable_if_t<is_iterable_v<typename NvpT::no_opt_type>, NvpT::array_element_type>>>
constexpr comparison_expr<NvpT, T> gt(
    const NvpT& lhs, const T& rhs) {
    return {lhs, rhs, "$gt"};
}
```